### DISCUSSION TITLE
An ergonomic "reverse identifier"

### DISCUSSION THREAD

**Discussion by amake** [petitparser/dart-petitparser#178] [Q&A] [answered]
The readme has the example of a parser that matches an identifier:

```dart
final id = letter() & (letter() | digit()).star();
```

I happened to need something equivalent to a *reverse* identifier: something that can have foos and bars up until the end, where it must end with a foo.

I tried to do something like

```dart
final id = (letter() | digit()).star() & letter();
```

But I understand that this doesn't work because there is no backtracking, correct?

My instinct was to use `starLazy` but the best limit condition I could come up with is pretty gnarly:

```dart
    final id =
        (letter() | digit()).starLazy(letter() & ((letter() | digit()).not())) &
            letter();
```

Is there an idiomatic or egonomic way to write this? I did come up with the following, but I always feel like `where` is cheating :)

```dart
    final id = [letter(), digit()]
        .toChoiceParser()
        .plus()
        .where((values) => letter().accept(values.last));
```

(Using `toChoiceParser` instead of `|` so that the type of `values.last` is known.)

**Comment by renggli** [comment, marked as answer]
Your example with `starLazy` is what I would have suggested. Only thing I would do is to extract a helper method, to make it more readable, i.e. 

```dart
Parser reverseId(Parser inner, Parser terminator) => 
    inner.starLazy(terminator & inner.not()) & terminator;

final id = reverseId(letter() | digit(), letter());
```

**Reply by amake** [reply]
Thanks, the refactor does make it a lot more readable!

**Comment by amake** [comment]
I have a new twist on this: Is it possible to match a "reverse ID" up to a trailing "inner" character? In other words:

- Reverse ID: 123a
- Reverse ID with trailing inner character: 123a0 (desire to match `123a` and let `0` be matched by another, non-reverse-ID parser)

I am reimplementing a feature that was originally using regex and this case was handled by backtracking. I can't see a way to fix this without breaking other edge cases.

**Reply by amake** [reply]
(This is pretty abstract as written; if it would help I can share my real use case, which is not really a "reverse ID" but is isomorphic to it.)

**Reply by renggli** [reply]
Yeah, I am a bit confused with the non-reverse-ID parser. Maybe if you could share the regex, and/or provide some more examples of what inputs are accepted/rejected?
